common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -39,17 +86,735 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+                oAuth2Strategy = oAuth2Strategy,
+                parametersWithScopes = parametersWithScopes, // Include the scopes, to send to /token endpoint in case of password flow
+                usePassword = usePassword
+            )
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signInStart", e)
+            throw e
+        }
+    }
+
+    /**
+     * Makes a call to the /token endpoint with the provided SLT, and caches the returned token if successful.
+     */
+    fun signInWithSLT(parameters: SignInWithSLTCommandParameters): SignInWithSLTCommandResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signInWithSLT")
+
+        try {
+            val oAuth2Strategy = createOAuth2Strategy(parameters)
+
+            val mergedScopes = addDefaultScopes(parameters.scopes)
+            val parametersWithScopes = CommandUtil.createSignInWithSLTCommandParametersWithScopes(
+                parameters,
+                mergedScopes
+            )
+
+            val tokenApiResult = performSLTTokenRequest(
+                oAuth2Strategy = oAuth2Strategy,
+                parameters = parametersWithScopes
+            )
+
+            when (tokenApiResult) {
+                is SignInTokenApiResult.Success -> {
+                    return saveAndReturnTokens(
+                        oAuth2Strategy = oAuth2Strategy,
+                        parametersWithScopes = parametersWithScopes,
+                        tokenApiResult = tokenApiResult
+                    )
+                }
+                is SignInTokenApiResult.InvalidAuthenticationType,
+                is SignInTokenApiResult.MFARequired, is SignInTokenApiResult.CodeIncorrect,
+                is SignInTokenApiResult.UserNotFound, is SignInTokenApiResult.InvalidCredentials,
+                is SignInTokenApiResult.UnknownError -> {
+                    Logger.warn(
+                        TAG,
+                        "Unexpected result: $tokenApiResult"
+                    )
+                    tokenApiResult as ApiErrorResult
+
+                    return ICommandResult.UnknownError(
+                        error = tokenApiResult.error,
+                        errorDescription = "API returned unexpected result: $tokenApiResult",
+                        errorCodes = tokenApiResult.errorCodes
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signInWithSLT", e)
+            throw e
+        }
+    }
+
+    /**
+     * Makes a call to the /token endpoint with the provided code, and caches the returned token if successful.
+     */
+    fun signInSubmitCode(parameters: SignInSubmitCodeCommandParameters): SignInSubmitCodeCommandResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signInSubmitCode")
+
+        try {
+            // Add default scopes
+            val mergedScopes: List<String> = addDefaultScopes(parameters.scopes)
```

Comment:
Since we add default scopes in the start method, do we need to add them again here? Or does that only happen for password flow

Reply:
In theory it can be removed here, as it's added to the first method calls in the sign in flows (`signInStart()` and `signInWithSLT()`). 
Why I like having it here, is so we're less likely to introduce a bug in the future if/when changes to the sign in flow are made and for some reason it's removed/not present in the methods that start the sign in flow. So while it's not DRY, it's added "just to be sure".
Is that acceptable?

Reply:
Yes! Sounds great! additional safety is always good